### ISSUE TITLE
worker process exited on signal 9 occasionally in kong proxy pod resulting API calling failed

### ISSUE THREAD

**Issue by lihongbj** [Kong/kong#9649]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Kong version (`$ kong version`)

3.0 and 2.5 

### Current Behavior

In OCP 4.10 cluster, We deploy kong as API gateway. In kong pod, there are 2 containers,  one is kong 3.0 from https://github.com/Kong/docker-kong, and the other is ingress 2.5 from https://github.com/Kong/kubernetes-ingress-controller. When POST/GET metric-api via kong, most completed successfully but occasionally failed with HTTP RC 502. When checking kong  log for the failed HTTP POST, it will be found that a worker process here is `31`  `signal 17 (SIGCHLD) received from 31` and then `worker process 31 exited on signal 9`:

```log
10.254.12.49 - - [25/Oct/2022:12:39:39 +0000] "POST /myapp/metric-api/v1/metrics HTTP/1.1" 200 1 "-" "curl/7.29.0"
10.254.24.83 - - [25/Oct/2022:12:39:40 +0000] "POST /myapp/metric-api/v1/metrics HTTP/1.1" 200 1 "-" "curl/7.29.0"
2022/10/25 12:39:41 [notice] 1#0: signal 17 (SIGCHLD) received from 31
2022/10/25 12:39:41 [alert] 1#0: worker process 31 exited on signal 9
2022/10/25 12:39:41 [notice] 1#0: start worker process 33
2022/10/25 12:39:41 [notice] 1#0: signal 29 (SIGIO) received
2022/10/25 12:39:41 [notice] 33#0: *740012 [kong] init.lua:290 only worker #0 can manage, context: init_worker_by_lua*
2022/10/25 12:39:41 [notice] 33#0: *740013 [kong] init.lua:290 only worker #0 can manage, context: init_worker_by_lua*
10.254.24.83 - - [25/Oct/2022:12:39:42 +0000] "POST /myapp/metric-api/v1/metrics HTTP/1.1" 200 1 "-" "curl/7.29.0"
10.254.12.49 - - [25/Oct/2022:12:39:43 +0000] "POST /myapp/metric-api/v1/metrics HTTP/1.1" 200 1 "-" "curl/7.29.0"
```

The worker process number in kong is 2.
Not found OOM/OOMKill for kong pod.
Kong pod memory is 512M.

### Expected Behavior

Expecting all API calling is successful and no worker process exit resulting API service is not available.

**Comment by rainest** [comment]
Pod-level OOM statuses apparently only appear [if the system killed PID 1](https://medium.com/@reefland/tracking-down-invisible-oom-kills-in-kubernetes-192a3de33a60) in the container, which the worker isn't.

You'd need to search through the journal or other host system logs regardless to pin down the issue. AFAIK there aren't any cases where the gateway code (really, NGINX, as it's the code that manages the workers) will send SIGKILL to workers, so it's presumably coming from outside.

512M is, however, quite low for a production Kong container. I don't have the current recommendations on hand, but I've seen Kong instances consume upwards of 2G quite often. Do you have historical metrics showing the Pod memory usage?

**Comment by morningspace** [comment]
Thanks @rainest for your comments!

> really, NGINX, as it's the code that manages the workers

Yeah, actually we are suspecting this is coming from the NGINX main process where it will send SIGKILL to one worker for some reason, then start a new one after that. We haven't checked journal or system logs, which is a very good idea. We are going to setup an env to reproduce the issue and check the logs. This will help us understand what happens w/ the worker.

On the other hand, I am a bit confused, given that worker die and renew is common thing in real world, why Kong (or really NGINX) can not shift requests to other workers, to avoid service unavailable like returning 502. Is it an issue of Kong, or NGINX, or something could be solved at k8s level, e.g.: increase replicas for HA?

> 512M is, however, quite low for a production Kong container.

This is not the production env, but our testing env internally. If 512M is not a good start, maybe we can increase it.

**Comment by rainest** [comment]
> can not shift requests to other workers, to avoid service unavailable like returning 502. Is it an issue of Kong, or NGINX, or something could be solved at k8s level, e.g.: increase replicas for HA?

Beyond my knowledge of how the gateway internals work; someone from that team may be able to answer better. I can at least confidently say that additional Kubernetes replicas wouldn't help though. You should have multiple to handle Pod restarts and the like, but the availability of other pods will have no bearing on the internal state of its neighbors.

Workers generally shouldn't die outside of expected scenarios (e.g. if you run `kong reload`). Determining why the kernel is terminating and addressing that should be the best approach here.

**Comment by fffonion** [comment]
Checking with the host system's kernel log (`dmesg -T`) is a good idea, whether it's a OOM kill or a sefgault
it will leave some trace there. And it will be good for further investigation. Usually nginx won't send signal
9, you can safely assume it's from outside.

**Comment by morningspace** [comment]
Yeah, ultimately we found that's caused by OOM kill which indicates that we need to increase the memory for kong proxy.

**Comment by lihongbj** [comment]
Thanks @rainest and @fffonion for your quick and professional help. Since root cause is found, I would like to close this now.
